spec/app/ics-028-cross-chain-validation/technical_specification.md

```diff
@@ -510,18 +521,72 @@ function onChanOpenConfirm(
 ```typescript
 // CCF: Consumer Chain Function
 // implements the AppModule interface
-function InitGenesis(state: ConsumerGenesisState) {
+function InitGenesis(gs: ConsumerGenesisState): [ValidatorUpdate] {
+  // ValidateGenesis
+  abortSystemUnless(gs.initialValSet NOT empty)
+  if gs.newChain {
+    // for new consumer chains, the genesis state 
+    // - contains a valid providerClientState  
+    abortSystemUnless(gs.providerClientState != nil)
+    abortSystemUnless(gs.providerClientState.Valid() == true)
+    // - contains a valid providerConsensusState
+    abortSystemUnless(gs.providerConsensusState != nil)
+    abortSystemUnless(gs.providerConsensusState.Valid() == true)
+    // - has the providerClientId unspecified
+    abortSystemUnless(gs.providerClientId == "")
+    // - has the providerChannelId unspecified
+    abortSystemUnless(gs.providerChannelId == "")
+    // - has the unbondingPackets empty
+    abortSystemUnless(gs.unbondingPackets IS empty)
+    // - contains the initial validator set that matches 
+    //   the validator set in the providerConsensusState (see ICS 7)
+    abortSystemUnless(gs.initialValSet == gs.providerConsensusState.validatorSet)
```

Comment:
We check hashes technically

Reply:
Yes, but this is a detail of implementation. ICS7 defines the ConsensusState as having a validator set, not a hash, i.e., https://github.com/cosmos/ibc/blob/marius%2Fccv-evidence/spec/client/ics-007-tendermint-client/README.md#consensus-state